utils/service_discovery/abstract_config_store.py

```diff
@@ -34,6 +35,141 @@ class KeyNotFound(Exception):
+
+    def _populate_auto_conf(self):
+        """Retrieve auto_conf templates"""
+        raw_templates = get_auto_conf_images(full_tpl=True)
+        for image, tpls in raw_templates.iteritems():
+            for check_name, init_tpl, instance_tpl in zip(*tpls):
+                if image in self.auto_conf_templates:
+                    if check_name in self.auto_conf_templates[image][0]:
+                        log.warning("Conflicting templates in auto_conf for image %s and check %s. "
+                                "Please check your template files." % (image, check_name))
+                        continue
+                    self.auto_conf_templates[image][0].append(check_name)
+                    self.auto_conf_templates[image][1].append(init_tpl)
+                    self.auto_conf_templates[image][2].append(instance_tpl)
+                else:
+                    self.auto_conf_templates[image][0] = [check_name]
+                    self.auto_conf_templates[image][1] = [init_tpl or {}]
+                    # no list wrapping because auto_conf files already have a list of instances
+                    self.auto_conf_templates[image][2] = instance_tpl or [{}]
+
+    def _issue_read(self, identifier):
+        """Perform a read against the KV store"""
+
+        # templates from the config store
+        try:
+            check_names = json.loads(
+                self.read_func(path.join(self.root_path, identifier, CHECK_NAMES).lstrip('/')))
+            init_config_tpls = json.loads(
+                self.read_func(path.join(self.root_path, identifier, INIT_CONFIGS).lstrip('/')))
+            instance_tpls = json.loads(
+                self.read_func(path.join(self.root_path, identifier, INSTANCES).lstrip('/')))
+            return [check_names, init_config_tpls, instance_tpls]
+        except KeyNotFound:
+            return None
+
+    def get_templates(self, identifier):
+        """
+        Return a dict of templates coming from the config store and
+        the auto_conf folder and their source for a given identifier.
+        Templates from kv_templates take precedence.
+        """
+        templates = {
+            # source: [[check_names], [init_configs], [instances]]
+            CONFIG_FROM_TEMPLATE: None,
+            CONFIG_FROM_AUTOCONF: None
+        }
+
+        # cache miss
+        if identifier not in self.kv_templates:
+            try:
+                tpls = self._issue_read(identifier)
+            except NotImplementedError:
+                # expected when get_check_names is called in auto-conf mode
+                tpls = None
+            except Exception:
+                tpls = None
+                log.exception('Failed to retrieve a template for %s.' % identifier)
+            # create a key in the cache even if _issue_read doesn't return a tpl
+            # so that subsequent reads don't trigger issue_read
+            self.kv_templates[identifier] = tpls or None
```

Comment:
I'm not sure you need the `or None`? `tpls` will already be something if `self._issue_read(identifier)` returns either a list or `None`, or the `None` we've set it in the `except` blocks.